Implementation of DifferentialExtension for 'tan'

gxyd opened the pull request:
Construction of DifferentialExtension of `tan`

- [x] Can handle Differential Extension of `f = tan(x) + tan(2*x)`

This task is similar to that of for the `exponentials`. But there are other ways in which DifferentialExtension of tangents could be algebraically dependent.

- [ ] Add methods to handle other ways in which algebraic dependence between tangents could be established.

There is a paper "Simplification of Real Elementary Functions", that paper gives some insight into dependency of tangents. It uses Real Structure Theorem.

- [x] Can handle DifferentialExtension of `f = tan(atan(x).rewrite(log))`

The above posed problem because of auto-simplification of `tan(I*t0)` to `I*tanh(t0)`. Aaron mentioned this issue in the comments.

gxyd committed:
Implementation of DifferentialExtension for 'tan'

gxyd committed:
Correct the "derivation" of 'tan'

jksuom commented on sympy/integrals/risch.py:
```diff
@@ -237,12 +238,14 @@ def __init__(self, f=None, x=None, handle_first='log', dummy=False, extension=No
         logs = set()
         symlogs = set()
+        tans = set()
+        symtans = set()
```
These sets could probably be internal to `_rewrite_tans` where functions with rationally related arguments are represented in terms of a single one.

gxyd commented on sympy/integrals/risch.py:
```diff
@@ -711,6 +756,20 @@ def decrement_level(self):
+    s0 = 0
+    n = as_int(n)
+    for k in range(1, (n + 1)//2 + 1):
+        s0 += nC(n, 2*k-1)*t**(2*k - 1)*(-1)**(k + 1)
```
`nC` doesn't work with sympified input. Is that a bug? I had to change class of `n` to `int`.

jksuom replied:
It is probably intentional, but I see no reason why it could not accept Integers.
There is `range(n + 1)` in the code, but n could easily be converted to int.

gxyd replied:
BTW, just to be clear that isn't an issue with `k` in `range` but with `n`, since that will still be `sympified` without use of `as_int`.

asmeurer commented on sympy/integrals/risch.py:
```diff
@@ -711,6 +756,20 @@ def decrement_level(self):
         self.case = self.cases[self.level]
         return None
 
+def rewrite_tans(n, t):
```
Add a docstring to this to explain what it does.

asmeurer commented on sympy/integrals/risch.py:
```diff
@@ -433,6 +443,41 @@ def _auto_attrs(self):
+        for arg, others in ip:
+            others.sort(key=lambda i: i[1])
+
+            arga, argd = frac_in(arg, self.t)
```
Was stepping through this in a debugger. These are not rewritten in terms of the extensions so far (like if you have `tan(log(x))` then `arg` will be `log(x)` instead of `t0`). Is that correct?

gxyd replied:
No, that is not correct, I think. May be, I understand what is wrong here.

gxyd replied:
I think this has been corrected now in the latest commit.

asmeurer commented:
Just a heads up, it looks like you might run into issues with things like

```
>>> tan(I*x)
I*tanh(x)
```

Maybe it would be better to replace all instances of `tan` with some dummy

```
class UnevaluatedTan(Function):
    pass
```

in the expression.

asmeurer commented:
This came up because I was trying to see what it did for `tan(atan(x).rewrite(log))`, an instance where I believe you will need the structure theorems. It ended up failing, but only because when it replaced the log with `t0`, the tan became a tanh, which confused it, and caused it to bail.

gxyd committed:
Add docstring for 'rewrite_tans'

gxyd committed:
Fix DifferentialExtension of composition of 'tan' with 'log'/'exp'

Also add tests for DifferentialExtension construction

gxyd committed:
Use "UnevaluatedTan" to avoid auto-simplification of tan(I*x)

Having the below auto-simplification can create problems

```
>>> tan(I*x)
I*tanh(x)
```

So add a new class 'UnevaluatedTan'

jksuom commented on sympy/integrals/risch.py:
```diff
@@ -761,6 +764,29 @@ def decrement_level(self):
+class UnevaluatedTan(Function):
+    """
+    Used to avoid auto-simplification of tangents of the type
+    'tan(I*x)'.
```
There is no need to avoid this kind of auto-simplification. When there is an `I` in the expression, real type monomial extensions (`tan`, `atan`) should not be used. In this case, `tanh(x)` is a rational expression of `exp(x)`, so an `exp` monomial can be used to compute the integral.

asmeurer replied:
But this needs to be done or else you will get incorrect results, e.g., using `tan(x).rewrite(exp) + tan(x)`.  

I don't know if we should be worried about other forms of complex constants (like `root(-1, 3)`). I guess if a constant is numeric we can always test if it is nonreal, and if it is symbolic, we can implicitly assume that it is real.

jksuom replied:
> tan(x).rewrite(exp) + tan(x)

I don't see how this could lead to incorrect results. If `tan(x)` was rewritten in (a part of) the input expression, then that would contain `I` and  `rewrite_complex` would be True. And then all the rest of the trigonometric functions would also be rewritten.

The implicit complex constants should be taken into account at some point, but I think that `I` could suffice for now. That is obviously the most common one.

asmeurer replied:
Sorry to clarify, I mean the `I` rewriting needs to be done or you will get wrong results. (right now it's in a separate PR)

gxyd committed:
Remove "UnevaluatedTan"

gxyd commented:
I have separated that change for any expression containing `I` into a separate PR https://github.com/sympy/sympy/pull/13225

gxyd committed:
'UnevaluatedTan' removal from test file

gxyd committed:
shift 'rewrite' logic to '_rewrite_tans'

gxyd committed:
modify 'tan' rewrite

gxyd committed:
Make final edit in rewrite of 'tan'

gxyd committed:
Merge branch 'master' into tans

Conflicts:
	sympy/integrals/risch.py

gxyd committed:
initalise implementation of paper 'Simplfiicaton of Real Elementary Function'

RituRajSingh878 commented:
I want to continue this work. So please help me.
I am unable to find  **Simplification of Real Elementary Functions** research paper on the internet.

So @gxyd if you have any source to download it, please tell me.

jksuom commented:
I did find this free link: http://percivalgopan7351.ml/cb57c9/simplification-of-real-elementary-functions-by-bronstein.pdf. I don't know if it will work. There is also a non-free link to ACM library.

czgdp1807 commented:
@RituRajSingh878 Have you continued this PR in an another one. Please provide the link to your PR or refer this one as `Closes #13095` in your PR. Thanks.
